bpf/bpf_lxc.c

```diff
@@ -53,6 +53,32 @@ static __always_inline bool redirect_to_proxy(int verdict, __u8 dir)
 }
 #endif
 
+#ifdef ENABLE_CUSTOM_CALLS
+/* Encode return value and identity into cb buffer. This is used before
+ * executing tail calls to custom programs. "ret" is the return value supposed
+ * to be returned to the kernel, needed by the callee to preserve the datapath
+ * logics. The "identity" reflects the local endpoint: the source of the packet
+ * on ingress path, or its destination on the egress path. We encode it so that
+ * custom programs can retrieve it and use it at their convenience.
+ */
+static __always_inline int
+encode_custom_prog_meta(struct __ctx_buff *ctx, int ret, __u32 identity)
+{
+	__u32 custom_meta = 0;
+
+	/* If we cannot encode return value on 8 bits, return an error so we can
+	 * skip the tail call entirely, as custom program has no way to return
+	 * expected value and datapath logics will break.
+	 */
+	if ((ret & 0xff) != ret)
+		return -1;
+	custom_meta |= (ret & 0xff) << 24;
+	custom_meta |= (identity & 0xffffff);
```

Comment:
Is this specifically trying to mirror `set_identity_mark()` in terms of bit layout? Would it make sense to either make a reference or refactor to reuse that code specifically to ensure this? (I realize this is putting the result into `skb->cb[]` instead of `skb->mark` though).

Reply:
It does _not_ attempt to mirror that layout. From what I understand, setting/retrieving the identity from and to the `skb->mark` needs this specific setup because 8 of the 24 bits were added [as an afterthought](https://github.com/cilium/cilium/pull/4858/files), and could not be placed next to the other 16 ones in the field.

No such constraint here: Just take the full 24 bits at once.
Before that we use the `ret` return value, because we need it in the custom program, but this is unrelated to what `set_identity_mark()` does.

So I do not believe there is much to refactor here.

Reply:
Ah yes, I saw the pattern and assumed it was the same but now I look again, it's clearly just putting the bits all together.